Lib/fontTools/ttLib/tables/otTables.py

```diff
@@ -1297,6 +1298,144 @@ def preWrite(self, font):
 		return self.__dict__.copy()
 
 
+class ClipBox(getFormatSwitchingBaseTableClass("uint8")):
+
+	def as_tuple(self):
+		return tuple(getattr(self, conv.name) for conv in self.getConverters())
+
+	def __repr__(self):
+		return f"{self.__class__.__name__}{self.as_tuple()}"
+
+
+class ClipList(BaseTable):
+
+	def populateDefaults(self, propagator=None):
+		if not hasattr(self, "clips"):
+			self.clips = {}
+
+	def postRead(self, rawTable, font):
+		clips = {}
+		glyphOrder = font.getGlyphOrder()
+		for i, rec in enumerate(rawTable["ClipRecord"]):
+			rangesOverlap = False
+			for glyphID in range(rec.StartGlyphID, rec.EndGlyphID + 1):
+				try:
+					glyph = glyphOrder[glyphID]
+				except IndexError:
+					continue
+				if glyph not in clips:
+					clips[glyph] = copy.copy(rec.ClipBox)
+				else:
+					rangesOverlap = True
+			if rangesOverlap:
+				log.warning(
+					"ClipRecord %i overlap previous records; "
+					"redefined clip boxes are skipped",
+					i,
+				)
+		self.clips = clips
+
+	def groups(self):
+		glyphsByClip = defaultdict(list)
+		uniqueClips = {}
+		for glyphName, clipBox in self.clips.items():
+			key = hash(clipBox.as_tuple())
```

Comment:
noob question, why is it necessary to call hash here rather than just using the tuple directly as the key?

Reply:
I think it's even wrong, as `hash()` is not guaranteed to be unique. `clipBox.as_tuple()` should be suitable as a dict key.

Reply:
Tuple is hashed by value so there should not be any difference. I thought hash() better conveyed the intent, but I'm ok to change to use a bare tuple (the hash function will be used internally by the dict type).

Reply:
A tuple of ints at least is hashed by value

Reply:
Typically hashes can overlap and you have to do a further search among the things with the same hash to establish a full match if there are >1 of them. I initially thought maybe this was some sort of Python trick but from Just comment it sounds like it works the normal way.

Reply:
> Tuple is hashed by value

Yes, but that still doesn't guarantee it's unique. The job of hash is to make collisions unlikely, but any collisions will be dealt with by dict or set.

Reply:
Got it, thanks for the clarification 👍